Feature/integrate offline queue

dron22 opened the pull request:
This PR integrates the offline queue into send heartbeats command. The following flags are added:

- `disable-offline`
- `disableoffline` (deprecated)
- `sync-offline-activity`

Alternatively `offline` boolean can be defined in `settings` section of `.wakatime.cfg` file. Default setting for number of heartbeats synced from offline queue is `100`, like in the python version.

gandarez commented on cmd/root.go:
```diff
@@ -17,6 +17,9 @@ const (
 	// defaultTimeoutSecs is the default timeout used for requests to the wakatime api
 	defaultTimeoutSecs = 60
+	// defaultOfflineSync is the default maximum number of heartbeats from the
+	// offline queue, which will be synced upon sending heartbeats to the API.
```
Ponctuation at the end is missing in the other comments. Should it be removed from this line?

dron22 replied:
Actually the linter should enforce punctuation at the end of comments. I suspect that the linter does not check comments on non-exported (lower-case) names. Nevertheless, let's keep punctuation and also add it to the other comments.

gandarez commented on pkg/offline/offline.go:
```diff
@@ -5,16 +5,40 @@ import (
+	if dir == "" {
+		dir, err = os.UserHomeDir()
+		if err != nil {
+			return "", fmt.Errorf("failed to retrieve user's home dir: %s", err)
```
just to follow the other messages like that it should be `failed getting user's home directory: %s`

dron22 replied:
Actually the other messages in the package are all in this format (`failed to ...`). Can you point me to where we use the other format?

gandarez commented on pkg/offline/offline.go:
```diff
@@ -5,16 +5,40 @@ import (
+
+	expanded, err := homedir.Expand(dir)
+	if err != nil {
+		return "", fmt.Errorf("failed to expand offline queue folder path: %s", err)
```
maybe `failed expanding offline queue folder path`?

dron22 commented on cmd/legacy/heartbeat/params.go:
```diff
@@ -123,6 +126,30 @@ func LoadParams(v *viper.Viper) (Params, error) {
 		isWrite = heartbeat.Bool(b)
 	}
 
+	offlineDisabled := vipertools.FirstNonEmptyBool(v, "disableoffline", "disable-offline")
```
This is the behavior of the python client. If you look at https://github.com/wakatime/wakatime/blob/master/wakatime/arguments.py#L340, you can see, that `args.offline` contains the `disable-offline` value. But `args.offline_deprecated` will overwrite it.

I agree that what you suggest is actually more appropriate. Should we ask @alanhamlett for an opinion here?

dron22 replied:
@alanhamlett In the python client, if both set, the deprecated `disableoffline` flag overwrites `disable-offline` flag.

What would you say regarding this edge case? Should we keep the behavior from the python version or adjust this?

alanhamlett replied:
I think for consistency let's keep the same logic as the Python cli. We'll start breaking with the Python cli in future releases, after making sure everything works the same way first.

dron22 commented:
@gandarez FYI, I added handling of heartbeat leftovers to the offline queue. This was missing in offline queue implementation and first I planned to push this in a follow up PR, but now I added it here.

dron22 commented:
FYI, the PR contains conflicting files, because filestats PR (#68) was merged recently, which also added flag params. Let's keep the commit history and conversation of the PR and finish the review/approval for the solution first. Then I'll rebase before merging it to master to remove the conflicts.
